src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,196 @@
+from __future__ import annotations
+
+import os
+
+from jinja2 import Environment, UndefinedError, nodes
+from packaging.version import Version
+from snowflake.cli.api.utils.dict_utils import deep_merge_dicts, deep_traverse
+from snowflake.cli.api.utils.graph import Graph, Node
+from snowflake.cli.api.utils.rendering import CONTEXT_KEY, get_snowflake_cli_jinja_env
+
+
+class Variable:
+    def __init__(self, vars_chain):
+        self._vars_chain = list(vars_chain)
+        self.templated_value = None
+        self.rendered_value = None
+
+    def get_key(self):
+        return ".".join(self._vars_chain)
+
+    def is_env_var(self):
+        return (
+            len(self._vars_chain) == 3
+            and self._vars_chain[0] == CONTEXT_KEY
+            and self._vars_chain[1] == "env"
+        )
+
+    def get_env_var_name(self) -> str:
+        if not self.is_env_var():
+            raise KeyError(
+                f"Referenced variable {self.get_key()} is not an environment variable"
+            )
+        return self._vars_chain[2]
+
+    def store_in_context(self, context: dict, value):
+        """
+        Takes a generic context dict to modify, and a value
+
+        Traverse through the multi-level dictionary to the location where this variables goes to.
+        Sets this location to the content of value.
+
+        Example: vars chain contains ['ctx', 'env', 'x'], and context is {}, and value is 'val'.
+        At the end of this call, context content will be: {'ctx': {'env': {'x': 'val'}}}
+        """
+        current_dict_level = context
+        for i, var in enumerate(self._vars_chain):
+            if i == len(self._vars_chain) - 1:
```

Comment:
```suggestion
            if i == chain_len:
```
Consider introducing variable